packages/scheduler/src/__tests__/Scheduler-test.internal.js

```diff
@@ -124,51 +167,311 @@ describe('Scheduler', () => {
     ]);
   });
 
-  it('prioritizes callbacks according to their timeouts', () => {
-    scheduleWork(() => doWork('A', 10), {timeout: 5000});
-    scheduleWork(() => doWork('B', 20), {timeout: 5000});
-    scheduleWork(() => doWork('C', 30), {timeout: 1000});
-    scheduleWork(() => doWork('D', 40), {timeout: 5000});
+  it('executes the highest priority callbacks first', () => {
+    scheduleCallback(() => doWork('A', 100));
+    scheduleCallback(() => doWork('B', 100));
 
-    // C should be first because it has the earliest timeout
-    expect(flushWork()).toEqual(['C', 'A', 'B', 'D']);
+    // Yield before B is flushed
+    expect(flushWork(100)).toEqual(['A']);
+
+    runWithPriority(InteractivePriority, () => {
+      scheduleCallback(() => doWork('C', 100));
+      scheduleCallback(() => doWork('D', 100));
+    });
+
+    // C and D should come first, because they are higher priority
+    expect(flushWork()).toEqual(['C', 'D', 'B']);
   });
 
-  it('times out work', () => {
-    scheduleWork(() => doWork('A', 100), {timeout: 5000});
-    scheduleWork(() => doWork('B', 200), {timeout: 5000});
-    scheduleWork(() => doWork('C', 300), {timeout: 1000});
-    scheduleWork(() => doWork('D', 400), {timeout: 5000});
+  it('expires work', () => {
+    scheduleCallback(() => doWork('A', 100));
+    runWithPriority(InteractivePriority, () => {
+      scheduleCallback(() => doWork('B', 100));
+    });
+    scheduleCallback(() => doWork('C', 100));
+    runWithPriority(InteractivePriority, () => {
+      scheduleCallback(() => doWork('D', 100));
+    });
 
-    // Advance time, but not by enough to flush any work
-    advanceTime(999);
+    // Advance time, but not by enough to expire any work
+    advanceTime(249);
     expect(clearYieldedValues()).toEqual([]);
 
-    // Advance by just a bit more to flush C
+    // Advance by just a bit more to expire the high pri callbacks
+    advanceTime(1);
+    expect(clearYieldedValues()).toEqual(['B', 'D']);
+
+    // Expire the rest
+    advanceTime(10000);
+    expect(clearYieldedValues()).toEqual(['A', 'C']);
+  });
+
+  it('has a default expiration of ~5 seconds', () => {
+    scheduleCallback(() => doWork('A', 100));
+
+    advanceTime(4999);
+    expect(clearYieldedValues()).toEqual([]);
+
+    advanceTime(1);
+    expect(clearYieldedValues()).toEqual(['A']);
+  });
+
+  it('continues working on same task after yielding', () => {
+    scheduleCallback(() => doWork('A', 100));
+    scheduleCallback(() => doWork('B', 100));
+
+    const tasks = [['C1', 100], ['C2', 100], ['C3', 100]];
+    const C = deadline => {
+      while (tasks.length > 0) {
+        doWork(...tasks.shift());
+        if (
+          tasks.length > 0 &&
+          !deadline.didTimeout &&
+          deadline.timeRemaining() <= 0
```

Comment:
Drive-by-comment: This seems to be the suggested scheduler usage code, so could you clarify these terms more? 1) Perhaps deadline.didTimeout is rather didExpire (as you have called these elsewhere), but is it a deadline that expires or is there a way to express this without negation. 2) Also what happens if one forgets to check expiration, I would guess the scheduler calls us immediately again so it works, just slower, or is the problem more drastic. 3) Also timeRemaining is about the current frame (or how do you call it) which might confuse the user as it reads now, just after the didTimeout.

Reply:
Thanks for the feedback. Note that the actual names used here are not final yet.

> Perhaps deadline.didTimeout is rather didExpire (as you have called these elsewhere), but is it a deadline that expires or is there a way to express this without negation

The `deadline` naming is inherited from `requestIdleCallback`. You're right that the naming is confusing because it's about the frame deadline, which is different from the expiration. We haven't figured out the best terms to use yet but we'll take all this into consideration before we reach stable.

> Also what happens if one forgets to check expiration, I would guess the scheduler calls us immediately again so it works, just slower, or is the problem more drastic.

If you forget to check if the work is expired, but you do check `timeRemaining()`, then it will still  work because `timeRemaining()` will be 0 (though I guess I'm missing a test for this). Maybe this implies that we should unify the two APIs into one (these are also inherited from `requestIdleCallback`). I think it's likely we'll replace both with a single `shouldYield()` method. But I do think it's useful to provide a estimate for how much time before the next frame. But that could be a separate API.